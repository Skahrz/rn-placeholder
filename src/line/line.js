import React from "react";
import { View } from "react-native";

export const Line = ({
  textSize = 12,
  color = "#efefef",
  style,
  noMargin = false,
  ...props
}) => {
  const height = textSize;
  const alignSelf = "stretch";
  const backgroundColor = color;
  const borderRadius = textSize / 4;
  const marginBottom = noMargin ? 0 : textSize;
  const computedStyle = {
    height,
    alignSelf,
    backgroundColor,
    borderRadius,
    marginBottom,
    marginHorizontal: 0
  };

  return <View style={[computedStyle, style]} {...props} />;
};
