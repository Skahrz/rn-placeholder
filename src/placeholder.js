import connect from './placeholderContainer';
import Paragraph from './paragraph/paragraph';
import Media from './media/media';
import Line from './line/line';
import ImageContent from './imageContent/imageContent';
import MultiWords from './multiWords/multiWords';
import Box from './box/box';

/**
 * Export the placeholder
 */
export default {
  ImageContent: connect(ImageContent),
  Paragraph: connect(Paragraph),
  Media: connect(Media),
  Line: connect(Line),
  MultiWords: connect(MultiWords),
  Box: connect(Box),
  connect,
};
