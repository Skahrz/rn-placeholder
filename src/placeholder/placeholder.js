import React from "react";
import { View, StyleSheet } from "react-native";
import Animations from "../animation/animations";

const styles = StyleSheet.create({
  container: {
    flexDirection: "row",
    flex: 1
  },
  centerElement: {
    flex: 1
  },
  leftSide: {
    marginRight: 12
  },
  rightSide: {
    marginLeft: 12
  }
});

const makeRoot = animation => {
  if (animation) {
    const Animation = Animations[animation];

    if (!Animation) {
      throw new Error(`Animation "${animation}" doesn't exist in the library`);
    }

    return Animation;
  }

  return View;
};

const withView = (fn, props) => <View {...props}>{fn()}</View>;

export const Placeholder = ({
  isReady,
  animation,
  customAnimation,
  children,
  whenReadyRender: WhenReadyRender,
  renderLeft,
  renderRight,
  useDefaultWidth = true,
  ...props
}) => {
  const Root = customAnimation || makeRoot(animation);

  if (isReady) {
    return WhenReadyRender ? <WhenReadyRender /> : null;
  }

  const childrenArray = React.Children.toArray(children);
  const sizeOfChildren = childrenArray.length;
  const containerWidth = useDefaultWidth ? { width: "100%" } : {};
  const userStyle = props.style;
  return (
    <Root {...props} style={[styles.container, userStyle, containerWidth]}>
      {renderLeft && withView(renderLeft, { style: styles.leftSide })}
      <View style={styles.centerElement}>
        {childrenArray.map((element, index) =>
          React.cloneElement(element, {
            noMargin: index === sizeOfChildren - 1
          })
        )}
      </View>
      {renderRight && withView(renderRight, { style: styles.rightSide })}
    </Root>
  );
};
