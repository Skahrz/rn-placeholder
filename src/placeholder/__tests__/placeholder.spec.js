import React from 'react';
import { View, Text } from 'react-native';
import { shallow } from 'enzyme';
import { Line } from '../../line/line';
import { Placeholder } from '../placeholder';

describe('placeholder', () => {
  let props;

  const getWrapper = () => shallow(
    <Placeholder {...props}>
      <Line />
      <Line />
      <Line />
    </Placeholder>,
  );

  beforeEach(() => {
    props = {
      isReady: false,
      whenReadyRender: () => <View testID="when-ready-render" />,
    };
  });

  it('should match snapshot', () => {
    expect(getWrapper()).toMatchSnapshot();
  });

  it('shouldnt display anything when whenReadyRender is undefined', () => {
    props.isReady = true;
    props.whenReadyRender = undefined;
    expect(getWrapper()).toMatchSnapshot();
  });

  it('should display the whenReadyRender when isReady is true', () => {
    props.isReady = true;
    expect(getWrapper()).toMatchSnapshot();
  });

  it('should be wrapped with an Animation when animate prop is set to "fade"', () => {
    props.animation = 'fade';

    expect(getWrapper()).toMatchSnapshot();
  });

  it('should throw an error when the animation name is not found', () => {
    props.animation = 'toto';

    expect(() => getWrapper()).toThrow('Animation "toto" doesn\'t exist in the library');
  });

  it('should be wrapped with a CustomAnimation when the customAnimation prop is set', () => {
    props.customAnimation = componentProps => (
      <View testID="custom-animation" {...componentProps} />
    );

    expect(getWrapper()).toMatchSnapshot();
  });

  it('should have a component on the left side when renderLeft is provided', () => {
    props.renderLeft = otherProps => <Text {...otherProps}>Left content</Text>;

    expect(getWrapper()).toMatchSnapshot();
  });

  it('should have a component on the right side when renderRight is provided', () => {
    props.renderRight = otherProps => <Text {...otherProps}>Right content</Text>;

    expect(getWrapper()).toMatchSnapshot();
  });
});
