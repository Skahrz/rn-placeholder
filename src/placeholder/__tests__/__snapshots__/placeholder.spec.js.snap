// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`placeholder should be wrapped with a CustomAnimation when the customAnimation prop is set 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "flexDirection": "row",
      "width": "100%",
    }
  }
  testID="custom-animation"
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "alignSelf": "stretch",
            "backgroundColor": "#efefef",
            "borderRadius": 3,
            "height": 12,
            "marginBottom": 12,
            "width": "100%",
          },
          undefined,
        ]
      }
    />
    <View
      style={
        Array [
          Object {
            "alignSelf": "stretch",
            "backgroundColor": "#efefef",
            "borderRadius": 3,
            "height": 12,
            "marginBottom": 12,
            "width": "100%",
          },
          undefined,
        ]
      }
    />
    <View
      style={
        Array [
          Object {
            "alignSelf": "stretch",
            "backgroundColor": "#efefef",
            "borderRadius": 3,
            "height": 12,
            "marginBottom": 0,
            "width": "100%",
          },
          undefined,
        ]
      }
    />
  </View>
</View>
`;

exports[`placeholder should be wrapped with an Animation when animate prop is set to "fade" 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "flexDirection": "row",
      "opacity": 0.5,
      "width": "100%",
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "alignSelf": "stretch",
            "backgroundColor": "#efefef",
            "borderRadius": 3,
            "height": 12,
            "marginBottom": 12,
            "width": "100%",
          },
          undefined,
        ]
      }
    />
    <View
      style={
        Array [
          Object {
            "alignSelf": "stretch",
            "backgroundColor": "#efefef",
            "borderRadius": 3,
            "height": 12,
            "marginBottom": 12,
            "width": "100%",
          },
          undefined,
        ]
      }
    />
    <View
      style={
        Array [
          Object {
            "alignSelf": "stretch",
            "backgroundColor": "#efefef",
            "borderRadius": 3,
            "height": 12,
            "marginBottom": 0,
            "width": "100%",
          },
          undefined,
        ]
      }
    />
  </View>
</View>
`;

exports[`placeholder should display the whenReadyRender when isReady is true 1`] = `
<View
  testID="when-ready-render"
/>
`;

exports[`placeholder should have a component on the left side when renderLeft is provided 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "flexDirection": "row",
      "width": "100%",
    }
  }
>
  <View
    style={
      Object {
        "marginRight": 12,
      }
    }
  >
    <Text>
      Left content
    </Text>
  </View>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "alignSelf": "stretch",
            "backgroundColor": "#efefef",
            "borderRadius": 3,
            "height": 12,
            "marginBottom": 12,
            "width": "100%",
          },
          undefined,
        ]
      }
    />
    <View
      style={
        Array [
          Object {
            "alignSelf": "stretch",
            "backgroundColor": "#efefef",
            "borderRadius": 3,
            "height": 12,
            "marginBottom": 12,
            "width": "100%",
          },
          undefined,
        ]
      }
    />
    <View
      style={
        Array [
          Object {
            "alignSelf": "stretch",
            "backgroundColor": "#efefef",
            "borderRadius": 3,
            "height": 12,
            "marginBottom": 0,
            "width": "100%",
          },
          undefined,
        ]
      }
    />
  </View>
</View>
`;

exports[`placeholder should have a component on the right side when renderRight is provided 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "flexDirection": "row",
      "width": "100%",
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "alignSelf": "stretch",
            "backgroundColor": "#efefef",
            "borderRadius": 3,
            "height": 12,
            "marginBottom": 12,
            "width": "100%",
          },
          undefined,
        ]
      }
    />
    <View
      style={
        Array [
          Object {
            "alignSelf": "stretch",
            "backgroundColor": "#efefef",
            "borderRadius": 3,
            "height": 12,
            "marginBottom": 12,
            "width": "100%",
          },
          undefined,
        ]
      }
    />
    <View
      style={
        Array [
          Object {
            "alignSelf": "stretch",
            "backgroundColor": "#efefef",
            "borderRadius": 3,
            "height": 12,
            "marginBottom": 0,
            "width": "100%",
          },
          undefined,
        ]
      }
    />
  </View>
  <View
    style={
      Object {
        "marginLeft": 12,
      }
    }
  >
    <Text>
      Right content
    </Text>
  </View>
</View>
`;

exports[`placeholder should match snapshot 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "flexDirection": "row",
      "width": "100%",
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "alignSelf": "stretch",
            "backgroundColor": "#efefef",
            "borderRadius": 3,
            "height": 12,
            "marginBottom": 12,
            "width": "100%",
          },
          undefined,
        ]
      }
    />
    <View
      style={
        Array [
          Object {
            "alignSelf": "stretch",
            "backgroundColor": "#efefef",
            "borderRadius": 3,
            "height": 12,
            "marginBottom": 12,
            "width": "100%",
          },
          undefined,
        ]
      }
    />
    <View
      style={
        Array [
          Object {
            "alignSelf": "stretch",
            "backgroundColor": "#efefef",
            "borderRadius": 3,
            "height": 12,
            "marginBottom": 0,
            "width": "100%",
          },
          undefined,
        ]
      }
    />
  </View>
</View>
`;

exports[`placeholder shouldnt display anything when whenReadyRender is undefined 1`] = `null`;
